### DISCUSSION TITLE
How to make a faster simulation

### DISCUSSION THREAD

**Discussion by Junang-Wang** [sofa-framework/sofa#4434] [BeamAdapter] [answered]
Hi there,

I'm currently working on a project that involves simulating the insertion of a catheter into a 2D Neurovascular model. The simulation is running quite slow, with the frames per second reducing to 13 at the end of the simulation. I'm wondering if anyone has any suggestions on how to improve the simulation. For example, would using a faster linear solver help? Or could the slow performance be due to using a Neurovascular model with a lot of redundant mesh which increase narrow collision detection steps in SOFA?
![image](https://github.com/sofa-framework/sofa/assets/89964840/96e6ccd7-0119-474f-87ee-a527d22c9825)
Thank you for your help in advance!

**Comment by Junang-Wang** [comment]
I attach my script below:
from plugin_list import pluginList, display
import numpy as np
def createScene(rootNode):

    # Header
    rootNode.addObject('FreeMotionAnimationLoop')
    rootNode.addObject('DefaultVisualManagerLoop')
    rootNode.addObject('LCPConstraintSolver',mu=0.1,tolerance=1e-4,maxIt=1000,build_lcp=False)
    rootNode.addObject('CollisionPipeline',draw=0,depth=6,verbose='1')
    rootNode.addObject('ParallelBruteForceBroadPhase',name='N2')
    rootNode.addObject('ParallelBVHNarrowPhase')
    rootNode.addObject('LocalMinDistance',contactDistance=1,alarmDistance=2,name='localmindistance',angleCone=0.5, coneFactor=0.5)
    rootNode.addObject('CollisionResponse',name='Response',response='FrictionContactConstraint')
    
    # setting VisualStyle, pluginList, and 3D frame coordinate
    rootNode.addObject('VisualStyle', displayFlags=display)
    rootNode.addObject('RequiredPlugin', pluginName=pluginList)
    rootNode.addObject('OglSceneFrame',style='Arrows', alignment='TopRight')
    rootNode.addObject('InteractiveCamera', name='Camera', position= [0,0,50], lookAt=[0,0,0])

    rootNode.gravity = [0.0, 0.0, 0.0] # mm/s^2
    radius = 1 #mm
    inner_radius = 0 
    StraightLength_stiff = 600 #mm
    StraightLength_soft = 400 #mm
    spireDiameter = 4000
    massDensity = 1550e-9 #kg/mm^3
    youngModulus_stiff = 10000#N/mm^2 = 10000e6Pa (stiff)
    youngModulus_soft = 10000
    poissonRatio = 0.3
    nbsection_soft = 10
    nbsection_stiff = 40

    
    #---------------------------------------------------------------------------
    # setting catheter topology mesh and wireRestShape
    GuideCatheter = rootNode.addChild('GuideCatheter')
    GuideCatheter.addObject('RodStraightSection', name ='StraightSection_stiff',
                         youngModulus= youngModulus_stiff, poissonRatio= poissonRatio, radius = radius, nbEdgesCollis = nbsection_stiff, nbEdgesVisu = 160, length = StraightLength_stiff)
    GuideCatheter.addObject('RodSpireSection', name ='StraightSection_soft',
                         youngModulus= youngModulus_soft, poissonRatio= poissonRatio, radius = radius, nbEdgesCollis = nbsection_soft, nbEdgesVisu = 40, length = StraightLength_soft, spireDiameter=spireDiameter, spireHeight=0)
    

    GuideCatheter.addObject('WireRestShape', name='GC_RestShape',  
                                 printLog=True, template='Rigid3d', wireMaterials = '@StraightSection_stiff @StraightSection_soft')
    GuideCatheter.addObject('EdgeSetTopologyContainer', name='meshLinesBeam')
    GuideCatheter.addObject('EdgeSetTopologyModifier', name='Modifier')
    GuideCatheter.addObject('EdgeSetGeometryAlgorithms', name='GeomAlgo', template='Rigid3d')
    GuideCatheter.addObject('MechanicalObject', name='dofTopo1', template='Rigid3d')
    


    #-----------------------------------------------------------------------
    # mechanical model of Catheter
    BeamModel = rootNode.addChild('BeamModel')
    
    BeamModel.addObject('EulerImplicitSolver', rayleighStiffness=0.2, printLog=False, rayleighMass=0.1)
    BeamModel.addObject('BTDLinearSolver', subpartSolve = False,verification = False, verbose=False)
    BeamModel.addObject('RegularGridTopology', name='MeshLines', drawEdges=False, 
                                    nx=180, ny=1, nz=1,
                                    xmax=1, xmin=0.0, ymin=0, ymax=0, zmax=1, zmin=1,
                                    p0=[0,0,0])
    BeamModel.addObject('MechanicalObject', showIndices=False, name='Instrument_DOFs', template='Rigid3d')
    BeamModel.addObject('WireBeamInterpolation', name='BeamInterpolation', WireRestShape='@../GuideCatheter/GC_RestShape', radius=radius, printLog=False)
    BeamModel.addObject('AdaptiveBeamForceFieldAndMass', name='BeamForceField', massDensity=massDensity, interpolation='@BeamInterpolation', computeMass=True, reinforceLength= False, shearStressComputation= False)

    #Deployment Controller
    BeamModel.addObject('InterventionalRadiologyController', name='DeployController', template='Rigid3d', instruments='BeamInterpolation', 
                                    startingPos=[0,0,0,0,0,0,1], xtip=[1, 0, 0], printLog=False, 
                                    rotationInstrument=[0, 0, 0], step=3, speed=2, 
                                    listening=True, controlledInstrument=0)
    BeamModel.addObject('LinearSolverConstraintCorrection', wire_optimization=True)

    BeamModel.addObject('FixedConstraint', indices=0, name='FixedConstraint')
    BeamModel.addObject('RestShapeSpringsForceField', name="RestSPForceField", points='@DeployController.indexFirstNode', angularStiffness=1e8, stiffness=1e8)
    BeamModel.addObject('PartialFixedConstraint', fixedDirections = [0,0,1,0,0,0],indices=np.arange(180))
    #-----------------------------------------------------------------------
    # Collision model
    Collis = BeamModel.addChild('CollisionCatheter')
    Collis.activated = True
    Collis.addObject('MechanicalObject', name='colli_DOFs',template='Vec3d')
    Collis.addObject('EdgeSetTopologyContainer',name='collisEdgeSet')
    Collis.addObject('EdgeSetTopologyModifier',name='colliseEdgeModifier')

    
    Collis.addObject('MultiAdaptiveBeamMapping',controller='../DeployController',useCurvAbs=True, printLog=False,name='collisMap')
    Collis.addObject('PointCollisionModel',proximity=0.0)
    Collis.addObject('LineCollisionModel', proximity=0.0)

    #--------------------------------------------------------------------
    # Visual model: using WireRestShape mesh
    VisualCatheter = BeamModel.addChild('VisualCatheter')
    
    VisualCatheter.addObject('MechanicalObject', name= 'visu_DOFs',template='Vec3d')
    
    VisualCatheter.addObject('QuadSetTopologyContainer', name = "visu_container")
    VisualCatheter.addObject('QuadSetTopologyModifier')
    VisualCatheter.addObject('QuadSetGeometryAlgorithms')
    # Mapping WireRestShape mesh to topology container
    VisualCatheter.addObject('Edge2QuadTopologicalMapping', nbPointsOnEachCircle=10,
                             radius=radius, input='@../../GuideCatheter/meshLinesBeam', output='@visu_container')
    # Mapping mesh to Beam Mechanics
    VisualCatheter.addObject('AdaptiveBeamMapping', input="@../Instrument_DOFs", output="@visu_DOFs", interpolation='@../BeamInterpolation')
    
    # Visual beam by OglModel
    VisuOgl = VisualCatheter.addChild('VisuOgl')
    VisuOgl.addObject("OglModel", name="visual", color=[0.5,1,0.5], quads="@../visu_container.quads")
    VisuOgl.addObject('IdentityMapping', input="@../visu_DOFs",output='@visual')

    

    #---------------------------------------------------------------------
    # Noticed that, flipNormals has to be set to 1 here, in order to set Env Collision Model collide from inner wall.
    Env= rootNode.addChild('Env')
    Env.addObject('MeshSTLLoader', name = 'env_mesh', filename= 'Neurovascular_2D_2.stl', translation=[0, 0, -1.0], rotation= [0.0, 0.0, 0.0],triangulate=True, scale = 1)
    Env.addObject('MeshTopology', position = '@env_mesh.position', triangles= '@env_mesh.triangles', drawTriangles=False)
    Env.addObject('MechanicalObject', name = 'env_DOFs', scale=3)
    
    Env.addObject('PointCollisionModel', moving=False, simulated = False)
    Env.addObject('TriangleCollisionModel', moving=False, simulated = False)
    Env.addObject('LineCollisionModel', moving= False, simulated = False)
    Env.addObject('OglModel', name='visu_env', src='@env_mesh', color=[1,0,0,0.3], scale=3)

    return rootNode

**Comment by alxbilger** [comment, marked as answer]
You need to profile the simulation to find out what the bottleneck is. Check out the following doc pages:
- https://www.sofa-framework.org/community/doc/using-sofa/performances/inspect-performances/
- https://www.sofa-framework.org/community/doc/using-sofa/performances/improve-performances/

According to the screenshot, your mesh for the vascular tree seems too much detailed. This could slow down the collision detection and add more constraints to solve. Try to simplify it, and to remove the unnecessary primitives.

**Reply by Junang-Wang** [reply]
Hi, alxbilger
Thank you for your response. It appears that you are correct. My mesh contains an excessive amount of details, which is causing the collision detection to take a significant amount of time(47% of total simulation time).  
![image](https://github.com/sofa-framework/sofa/assets/89964840/899c15fd-b13e-4b7d-bcfd-fa19cfeff531)
